### ISSUE TITLE
[Question] Merging top files for file_roots and pillar_roots

### ISSUE THREAD

**Issue by arizvisa** [saltstack/salt#51994]
### Description of Issue/Question
As per https://docs.saltstack.com/en/latest/ref/configuration/master.html#std:conf_master-file_roots, it seems that you can specify more than one root directory for both the roots and the pillars. However, it seems that the contents of these are not merged properly, as that the contents of only one of the topfiles takes priority.

Was this intentional, and if so...is there any way to merge the topfiles for pillars and roots within the same environment? I see there's a way to merge topfiles, but only for different environments which seems to be not possible if a minion is pinned to a specific environment.

### Setup
Specify more than one root directory for an environment. Better yet, do something like the following:
```
file_roots:
    base:
        - /srv/master
        - /srv/user
    master:
        - /srv/master
```

Then you can make a topfile like the following in `/srv/master/top.sls`:
```
base:
    '*':
        - state1
master:
    '*':
        - state2
```

Then something like the following in `/srv/user/top.sls`:
```
base:
    '*':
        - state2
```

The idea here is that that both `state1` and `state2` will be used for the `base` environment, and this way an individual user can modify `/srv/user/top.sls` without interfering with the `master` topfile at `/srv/master/top.sls`.

In essence, the final topfile should look like:
```
base:
    '*':
        - state1
        - state2
master:
    '*':
        - state2
```

### Steps to Reproduce Issue
When you call `state.show_top` on the minion (or whatever function used to show the pillar's topfile), it'll only only show the contents of one topfiles created. With the example I provided, if one tries to `state.show_top` for the `base` environment, it'll only show `state1` or `state2` rather than the combination of them.

It seems as if `salt.utils.dictupdate.update` is not being used to merge the topfiles between different topfiles, and instead a loop is being used to copy the contents of one topfile into the other..

### Versions Report
2019.2.0

**Comment by Ch3LL** [comment]
top files are merged across environments. It looks for a top.sls file in each environment and then depending on your `top_file_merging_strategy` it will merge those together.

For your use case what you will want is this:

```
file_roots:
    base:
        - /srv/user
    master:
        - /srv/master
top_file_merging_strategy: merge_all
```

This will merge the top files from both /srv/user and /srv/master the way you are desiring, because of the `merge_all` top file merging strategy. `merge_all` grabs all the states and top files from all environments and merges them together. 

The default `top_file_mergeing_strategy` is `merge`. The reason the default `merge` does not work for your use case is because by default the base's top.sls file is evaluated first, followed by other environments top files. The first target expression `*` it sees for a given environment it will use that one, and ignores the same target expression if found in other environments top files. The docs here: https://docs.saltstack.com/en/latest/ref/configuration/master.html#top-file-merging-strategy probably does a much better job at explain this as well :)

let me know if that helps :)

**Comment by arizvisa** [comment]
Yea, that makes sense. However, I have other environments as well and essentially want to merge just the same paths that are used for /srv/salt with /srv/master/salt without having the states in the base environment interfere with the other environment as I'm using environments to essentially define roles. (maybe this is wrong)

From what I understand, this isn't possible without having it merge across different environments as you described, right? And if not, is this something you guys would be willing to accept a PR for?

The use-case is that I'm using file permissions to restrict users from modifying states in a path/topfile so that one topfile takes priority for a specific environment (so a user can't break anything without having explicit perms).

So this way, there's the base environment which handles all remote minions and another environment which is used for containers that are built locally. I'd like to expose the ability for users to apply states to the container environment while still applying some default states to provision it. Then, the base environment will be applied to everything that is non-containerized.

Since a topfile allows you to specify more than one environment, merging the topfiles from multiple paths allows one to append to the specific environment without forcing a user to use targeting to isolate the minion states from one another.

**Comment by Ch3LL** [comment]
yes if you take a look at the code here: https://github.com/saltstack/salt/blob/v2019.2.0/salt/state.py#L3063

it gets all the environments and loops through each environment to look for a top file.

If you want to propose changing this you would need to go through our new salt enhancement proposal process. Here is the repo: https://github.com/saltstack/salt-enhancement-proposals

**Comment by arizvisa** [comment]
Awesome. Thanks for locating this code.

From looking at it, it looks if this logic was moved up into the roots, it would satisfy my needs as this is commented as if it intended to merge topfiles, when really it has the effect of merging environments since it's being done during state compilation (after roots have been figured out).  Unfortunately, since this is being done during state compilation being able to merge a topfile from multiple roots would likely require this complex logic, and the logic for the roots to be refactored a bit.

However, it looks like this shows that the roots are flattened in the configuration before it gets used by this code which is probably why it's difficult to merge multiple topfiles for the same root/environment and merging environments was chosen as a easier capability instead. So maybe multiple paths being able to be specified per root was more of an addition that was just easy to add rather than something that was planned from the very beginning which is why they just overwrite each other instead of merging like most of the other things in salt that I've fiddled with.

Anyways, hopefully I can analyze this properly and word it sanely in a proposal and get it accepted, heh.

I'm going to leave this ticket open for personal tracking. Are there any tags you have available to mark this as ongoing since this isn't really a question anymore?
